### PR TITLE
sdl_joypad: fix hotplug, order ports

### PR DESCRIPTION
**Pull request by Informatic**
## Description

Minor SDL joypad handling cleanup.

With joypad ignoring code used on webOS we were ending up with joypads registered at port 3+ on certain platforms. This also broke input port device index rebinding (Device Index list was always empty, even for devices on port 3+ which were working properly otherwise; not sure if missing reported devices is a violation of input device reporting API, or it's just a bug there)

This now properly attempts to store joypads continuously, by indexing them based on port number in an internal joypads state array.

Additionally, this fixes hotplug/unplug detection on SDL2. `event.jdevice.which` on unplug contains a Joystick Instance ID (monotonously increasing on every device connected/disconnected device), not Joystick Index used on open, thus a list of opened joysticks needs to be walked through and compared to received ID. Unplug detection would break after second replug before.

Note: this has not been tested outside of webOS, especially on SDL1.

## Related Issues

[Any issues this pull request may be addressing]

## Related Pull Requests

[Any other PRs from related repositories that might be needed for this pull request to work]

## Reviewers

[If possible @mention all the people that should review your pull request]

### CODE REVIEW

**Comment by HugLifeTiZ** [comment]
Is there a particular reason this PR has sat dormant? Hotplugging does not work on Steam Deck. Emudeck configures SDL as the gamepad backend because the normal user doesn't have the permissions necessary for the udev backend to work.

**Comment by hizzlekizzle** [comment]
@sonninnos @LibretroAdmin @Myaats Any thoughts? (including Mats due to Steam/Deck angle)

File: input/drivers_joypad/sdl_joypad.c [hunk 1]
```diff
@@ -55,10 +56,10 @@ static const char *sdl_joypad_name(unsigned pad)
 
 #ifdef HAVE_SDL2
    if (sdl_pads[pad].controller)
-      return SDL_GameControllerNameForIndex(pad);
-   return SDL_JoystickNameForIndex(pad);
+      return SDL_GameControllerNameForIndex(sdl_pads[pad].sdl_id);
```
Comment (RobLoach) [inline input/drivers_joypad/sdl_joypad.c:59]: Not really needed, but we could likely use [SDL_GameControllerName](https://wiki.libsdl.org/SDL2/SDL_GameControllerName) here.
``` suggestion
      return SDL_GameControllerName(sdl_pads[pad].controller);
```

File: input/drivers_joypad/sdl_joypad.c [hunk 2]
```diff
@@ -447,10 +475,20 @@ static void sdl_joypad_poll(void)
       switch (event.type)
       {
          case SDL_JOYDEVICEADDED:
+            RARCH_WARN("[SDL_JOYPAD] hotplug added: %d\n", event.jdevice.which);
             sdl_pad_connect(event.jdevice.which);
             break;
          case SDL_JOYDEVICEREMOVED:
```
Comment (RobLoach) [inline input/drivers_joypad/sdl_joypad.c:481]: Should we consider implementing [SDL_ControllerDeviceEvent](https://wiki.libsdl.org/SDL2/SDL_ControllerDeviceEvent)s `SDL_CONTROLLERDEVICEADDED` and `SDL_CONTROLLERDEVICEREMOVED`?

**Review by RobLoach — APPROVED** [review]
Looks like a solid change and clean up to SDL's gamepad control system.

**Comment by sonninnos** [comment]
Most of those loggings don't really seem like "warnings" but rather regular "info"..

**Comment by HugLifeTiZ** [comment]
Thanks for the review, @RobLoach. I would love to see this merged in time for the next release.

File: input/drivers_joypad/sdl_joypad.c [hunk 3]
```diff
@@ -91,9 +92,27 @@ static int16_t sdl_pad_get_axis(sdl_joypad_t *pad, unsigned axis)
...
 
+// id is a SDL-side id
 static void sdl_pad_connect(unsigned id)
 {
-   sdl_joypad_t *pad          = (sdl_joypad_t*)&sdl_pads[id];
+   int port_num = -1;
+
+   for (int i = 0; i < MAX_USERS; i++) {
+       if (sdl_pads[i].joypad != NULL && sdl_pads[i].sdl_id == id) {
+           RARCH_WARN("[SDL_JOYPAD]: %d already registered on port %d\n", id, i);
```
Comment (RobLoach) [inline input/drivers_joypad/sdl_joypad.c:102]:
``` suggestion
           RARCH_INFO("[SDL_JOYPAD]: %d already registered on port %d\n", id, i);
```

File: input/drivers_joypad/sdl_joypad.c [hunk 4]
```diff
@@ -91,9 +92,27 @@ static int16_t sdl_pad_get_axis(sdl_joypad_t *pad, unsigned axis)
...
+           return;
+       }
+   }
+
+   for (int i = 0; i < MAX_USERS; i++) {
+       if (sdl_pads[i].joypad == NULL) {
+           RARCH_WARN("[SDL_JOYPAD]: using port: %d\n", i);
```
Comment (RobLoach) [inline input/drivers_joypad/sdl_joypad.c:109]:
``` suggestion
           RARCH_LOG("[SDL_JOYPAD]: using port: %d\n", i);
```

File: input/drivers_joypad/sdl_joypad.c [hunk 5]
```diff
@@ -104,6 +123,7 @@ static void sdl_pad_connect(unsigned id)
 
    if (SDL_IsGameController(id))
    {
+      RARCH_WARN("[SDL_JOYSTICK]: opening #%d as GameController\n", id);
```
Comment (RobLoach) [inline input/drivers_joypad/sdl_joypad.c:126]:
``` suggestion
      RARCH_LOG("[SDL_JOYSTICK]: opening #%d as GameController\n", id);
```

File: input/drivers_joypad/sdl_joypad.c [hunk 6]
```diff
@@ -112,6 +132,7 @@ static void sdl_pad_connect(unsigned id)
    else
 #endif
    {
+      RARCH_WARN("[SDL_JOYSTICK]: opening #%d as Joystick\n", id);
```
Comment (RobLoach) [inline input/drivers_joypad/sdl_joypad.c:135]:
``` suggestion
      RARCH_LOG("[SDL_JOYSTICK]: opening #%d as Joystick\n", id);
```

File: input/drivers_joypad/sdl_joypad.c [hunk 7]
```diff
@@ -139,23 +161,29 @@ static void sdl_pad_connect(unsigned id)
    product    = guid_ptr[1];
 #endif
 #ifdef WEBOS
+   RARCH_WARN("[SDL_JOYPAD] #%d: vendor: %04x; product: %04x\n", id, vendor, product);
    if (vendor == 0x9999 && product == 0x9999)
    {
-      RARCH_WARN("[SDL_JOYPAD]: Ignoring pad #%d (vendor: %d; product: %d)\n", id, vendor, product);
-      if (pad->joypad)
-         SDL_JoystickClose(pad->joypad);
+      RARCH_WARN("[SDL_JOYPAD]: Ignoring pad #%d (vendor: %04x; product: %04x)\n", id, vendor, product);
```
Comment (RobLoach) [inline input/drivers_joypad/sdl_joypad.c:167]:
``` suggestion
      RARCH_LOG("[SDL_JOYPAD]: Ignoring pad #%d (vendor: %04x; product: %04x)\n", id, vendor, product);
```

File: input/drivers_joypad/sdl_joypad.c [hunk 8]
```diff
@@ -139,23 +161,29 @@ static void sdl_pad_connect(unsigned id)
...
       pad->joypad = NULL;
+      pad->controller = NULL;
       return;
    }
 #endif
 #endif
 
+   pad->sdl_id = id;
+
+   RARCH_WARN("[SDL_JOYPAD]: registering %d as port %d\n", id, port_num);
```
Comment (RobLoach) [inline input/drivers_joypad/sdl_joypad.c:180]:
``` suggestion
   RARCH_LOG("[SDL_JOYPAD]: registering %d as port %d\n", id, port_num);
```

File: input/drivers_joypad/sdl_joypad.c [hunk 9]
```diff
@@ -447,10 +475,20 @@ static void sdl_joypad_poll(void)
...
+            RARCH_WARN("[SDL_JOYPAD] hotplug added: %d\n", event.jdevice.which);
             sdl_pad_connect(event.jdevice.which);
             break;
          case SDL_JOYDEVICEREMOVED:
-            sdl_pad_disconnect(event.jdevice.which);
+            RARCH_WARN("[SDL_JOYPAD] hotplug removed: %d\n", event.jdevice.which);
+
+            for (int i = 0 ; i < MAX_USERS ; i++) {
+                if (SDL_JoystickInstanceID(sdl_pads[i].joypad) == event.jdevice.which) {
+                    RARCH_WARN("Found joystick: %d\n", i);
```
Comment (RobLoach) [inline input/drivers_joypad/sdl_joypad.c:486]:
``` suggestion
                    RARCH_LOG("[SDL_JOYPAD] Found joystick: %d\n", i);
```

**Comment by gouchi** [comment]
@TiZ-HugLife If you can resolve latest suggested changes.

Thank you.

**Comment by RobLoach** [comment]
The changes represented here have been brought into https://github.com/libretro/RetroArch/pull/16941 .

**Pull request closed without merging**